### PR TITLE
CORS_support_for_events

### PR DESCRIPTION
**Pull request by goru97**
What:
This PR solves two changes:
1. Support for multiple methods (GET, POST, OPTIONS etc) for same URL.
2. Support for OPTIONS method for events endpoint, which is used by the browser as preflight query before sending actual GET request with Authorization token.

Why:
1. Blueflood plugin for Grafana is under-construction that will run in browser's engine and will directly contact blueflood. 
2. These changes were required to support Cross Origin Resource Sharing with blueflood.

How: 
1. Fixed RouteMatcher class.
2. Added options method in HttpMetricDataQueryServer class.

### CODE REVIEW

File: blueflood-http/src/main/java/com/rackspacecloud/blueflood/outputs/handlers/HttpEventsQueryHandler.java [hunk 1]
```diff
@@ -27,35 +28,46 @@ public HttpEventsQueryHandler(EventsIO searchIO) {
...
+            }
+        }
+        else if(request.getMethod().toString().equalsIgnoreCase("OPTIONS")){
+            HttpResponse response = new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK);
+            response.setHeader("Access-Control-Allow-Origin", "*");
+            response.setHeader("Access-Control-Allow-Methods", "GET");
+            response.setHeader("Access-Control-Allow-Headers", "X-Auth-Token, Accept");
+            response.setHeader("Access-Control-Max-Age", "1728000");
+            HttpResponder.respond(ctx, request, response);
         }
```
Comment (VinnyQ) [inline blueflood-http/src/main/java/com/rackspacecloud/blueflood/outputs/handlers/HttpEventsQueryHandler.java]: Do we need a final else case here to handle where the request is neither GET or OPTIONS?

Comment (VinnyQ) [inline blueflood-http/src/main/java/com/rackspacecloud/blueflood/outputs/handlers/HttpEventsQueryHandler.java]: Oh never mind, this class is only currently invoked in either of those two cases.  It's cool no need for a catchall else.

File: blueflood-http/src/main/java/com/rackspacecloud/blueflood/outputs/handlers/HttpEventsQueryHandler.java [hunk 2]
```diff
@@ -27,35 +28,46 @@ public HttpEventsQueryHandler(EventsIO searchIO) {
...
     public void handle(ChannelHandlerContext ctx, HttpRequest request) {
-        final String tenantId = request.getHeader("tenantId");
-        HttpResponseStatus status = HttpResponseStatus.OK;
 
-        ObjectMapper objectMapper = new ObjectMapper();
-        String responseBody = null;
-        final Timer.Context httpEventsFetchTimerContext = httpEventsFetchTimer.time();
-        try {
-            HTTPRequestWithDecodedQueryParams requestWithParams = (HTTPRequestWithDecodedQueryParams) request;
-            Map<String, List<String>> params = requestWithParams.getQueryParams();
```
Comment (VinnyQ) [inline blueflood-http/src/main/java/com/rackspacecloud/blueflood/outputs/handlers/HttpEventsQueryHandler.java]: Not sure how it got missed here, but can you add the line "Tracker.track(request);" here for debugging issues in production via jolokia/jmx?  See HttpMetricsIngestionHandler.handle() method as an example.  Thanks!

Comment (tilogaat) [inline blueflood-http/src/main/java/com/rackspacecloud/blueflood/outputs/handlers/HttpEventsQueryHandler.java]: good point @VinnyQ 
I think we added the annotations endpoint at the same time as you were adding the jmx stuff, which is why it got left out.

Comment (goru97) [inline blueflood-http/src/main/java/com/rackspacecloud/blueflood/outputs/handlers/HttpEventsQueryHandler.java]: The response is being sent by DefaultHandler for the get request; There is "Tracker.trackResponse(request, response);" over there. I'm not sure if you still want this?

File: blueflood-http/src/main/java/com/rackspacecloud/blueflood/outputs/handlers/HttpEventsQueryHandler.java (hunk 1 above)
Comment (tilogaat) [inline blueflood-http/src/main/java/com/rackspacecloud/blueflood/outputs/handlers/HttpEventsQueryHandler.java]: Can we check if the access-control-allow-origin and all other headers can be added to repose?

**Comment by tilogaat** [comment]
+1 after the repose question is answered

**Pull request merged**